clippy_lints/src/needless_bool.rs

```diff
@@ -161,7 +161,9 @@ impl<'tcx> LateLintPass<'tcx> for NeedlessBool {
                     snip = snip.blockify();
                 }
 
-                if condition_needs_parentheses(cond) && is_parent_stmt(cx, e.hir_id) {
+                if (condition_needs_parentheses(cond) && is_parent_stmt(cx, e.hir_id))
+                    || is_receiver_of_method_call(cx, e)
+                {
                     snip = snip.maybe_par();
```

Comment:
Most lints just unconditionally call `maybe_par` since that can already figure out on its own if it needs parentheses, I wonder why we aren't just doing this here? Would this otherwise add too many unnecessary parens?

FWIW, method call receivers are not the only thing where we'd need parentheses. `&&` has very low precedence so they are needed for almost any parent expression (e.g. any binary expression other than `||`).

```rs
let (a, b) = Default::default();

_ = if a && b { true } else { false } as u8; // should also suggest `(a && b) as u8`
```

Reply:
> Most lints just unconditionally call `maybe_par` since that can already figure out on its own if it needs parentheses, I wonder why we aren't just doing this here? Would this otherwise add too many unnecessary parens?

Yup.

> FWIW, method call receivers are not the only thing where we'd need parentheses. `&&` has very low precedence so they are needed for almost any parent expression (e.g. any binary expression other than `||`).
> 
> ```rust
> let (a, b) = Default::default();
> 
> _ = if a && b { true } else { false } as u8; // should also suggest `(a && b) as u8`
> ```

I'll check this one as well, I'll put the PR in WIP mode in the meantime.